core/embed/rust/src/error.rs

```diff
@@ -1,46 +1,59 @@
-use core::convert::Infallible;
-use core::fmt;
-
+use core::convert::{Infallible, TryInto};
 use cstr_core::CStr;
 
+use crate::micropython::{ffi, obj::Obj, qstr::Qstr};
+
+#[derive(Debug)]
 pub enum Error {
-    Missing,
+    TypeError,
     OutOfRange,
-    InvalidType,
-    NotBuffer,
-    NotInt,
-    InvalidOperation,
+    MissingKwargs,
+    AllocationFailed,
+    CaughtException(Obj),
+    KeyError(Obj),
+    AttributeError(Qstr),
+    ValueError(&'static CStr),
+    ValueErrorParam(&'static CStr, Obj),
 }
```

Comment:
It's a bit unfortunate that this ties the `Error` type to MicroPython, and pollutes all the modules with it, but I suppose that a single error type is a bit of a (very convenient) anti-pattern anyway.

Reply:
true, but it turns out that we don't do basically any error handling within Rust and the Error type is almost always supposed to fall through to uPy